src/data/polynomial/erase_lead.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2020 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.polynomial.degree.basic
+import data.polynomial.degree.trailing_degree
+
+/-!
+# Erase the leading term of a univariate polynomial
+
+## Definition
+
+* `erase_lead f`: the polynomial `f - leading term of f`
+
+`erase_lead` serves as reduction step in an induction, shaving off one monomial from a polynomial.
+The definition is set up so that it does not mention subtraction in the definition,
+and thus works for polynomials over semirings as well as rings.
+-/
+
+noncomputable theory
+open_locale classical
+
+open polynomial finsupp finset
+
+namespace polynomial
+
+variables {R : Type*} [semiring R] {f : polynomial R}
+
+/-- `erase_lead f` for a polynomial `f` is the polynomial obtained by
+subtracting from `f` the leading term of `f`. -/
+def erase_lead (f : polynomial R) : polynomial R :=
+finsupp.erase f.nat_degree f
+
+section erase_lead
+
+lemma erase_lead_support (f : polynomial R) :
+  f.erase_lead.support = f.support.erase f.nat_degree :=
+by convert rfl
```

Comment:
Does `rfl` fail? Why?

Reply:
This lemma was proved by Johan.  I can confirm that ```rfl``` alone does not work, with error
```
type mismatch at application
  tactic.istep 43 3 43 3 rfl
term
  rfl
has type
  ?m_2 = ?m_2 : Prop
but is expected to have type
  tactic ?m_1 : Type ?
```

Reply:
`rfl`, not `by rfl`.

Reply:
Doesn't work, probably because of different `decidable` instances.

Reply:
It wouldn't hurt to mention this in a comment:
```suggestion
-- `rfl` doesn't work, probably because of different `decidable` instances
by convert rfl
```

Reply:
I don't understand why they're different: both sides use classical logic.

Reply:
Hmm. Then maybe just:
```suggestion
-- `rfl` doesn't work, TODO: why?
by convert rfl
```

Reply:
Polynomials use `nat.decidable_eq` which isn't classical. Hence `rfl` fails :cry:

Reply:
If the issue is indeed different `decidable_eq`s, could you add back @bryangingechen's first suggestion?